src/unix/pty.cc

```diff
@@ -131,6 +132,131 @@ pty_after_waitpid(uv_async_t *);
+
+  // temporarily block all signals
+  // this is needed due to a race condition in openpty
+  // and to avoid running signal handlers in the child
+  // before exec* happened
+  sigfillset(&newmask);
+  pthread_sigmask(SIG_SETMASK, &newmask, &oldmask);
+
+  pid = pty_forkpty(&pty_master, nullptr, term, winp);
+
+  if (!pid) {
+    // remove all signal handler from child
+    sig_action.sa_handler = SIG_DFL;
+    sig_action.sa_flags = 0;
+    sigemptyset(&sig_action.sa_mask);
+    for (int i = 0 ; i < NSIG ; i++) {    // NSIG is a macro for all signals + 1
+      sigaction(i, &sig_action, NULL);
+    }
+  }
+  // reenable signals
+  pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+
+  if (pid) {
+    if (pid == -1) {
+      Nan::ThrowError("forkpty(3) failed.");
+      return 1;
+    }
+
+    return 0;
+  } else {
+    if (strlen(cwd)) {
+      if (chdir(cwd) == -1) {
+        perror("chdir(2) failed.");
+        _exit(1);
+      }
+    }
+
+    if (uid != -1 && gid != -1) {
+      if (setgid(gid) == -1) {
+        perror("setgid(2) failed.");
+        _exit(1);
+      }
+      if (setuid(uid) == -1) {
+        perror("setuid(2) failed.");
+        _exit(1);
+      }
+    }
+
+    pty_execvpe(argv[0], argv, env);
+
+    perror("execvp(3) failed.");
+    _exit(1);
+  }
+  return 0;
+}
+
+int pty_fork_posix_spawn(char **argv, char **env, const termios *term, winsize *winp, int &pty_master, pid_t &pid) {
+  sigset_t all_signals, oldmask;
+
+  // temporarily block all signals
+  // this is needed due to a race condition in openpty
+  // and to avoid running signal handlers in the child
+  // before exec* happened
+  sigfillset(&all_signals);
+  pthread_sigmask(SIG_SETMASK, &all_signals, &oldmask);
+
+  int pty_slave;
+  if (pty_openpty(&pty_master, &pty_slave, nullptr, term, winp) == -1) {
+    pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+    Nan::ThrowError("pty_openpty failed.");
+    return 1;
+  }
+
+  pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+
+  posix_spawn_file_actions_t acts;
+  posix_spawnattr_t attrs;
+  if (posix_spawn_file_actions_init(&acts)) {
+    Nan::ThrowError("posix_spawn_file_actions_init failed.");
+    return 1;
+  }
+  if (posix_spawnattr_init(&attrs)) {
+    Nan::ThrowError("posix_spawnattr_init failed.");
+    return 1;
+  }
+
+  if (
+    posix_spawn_file_actions_adddup2(&acts, pty_slave, STDIN_FILENO) ||
+    posix_spawn_file_actions_adddup2(&acts, pty_slave, STDOUT_FILENO) ||
+    posix_spawn_file_actions_addclose(&acts, pty_slave) ||
+    posix_spawn_file_actions_addclose(&acts, pty_master)
+  ) {
+    posix_spawn_file_actions_destroy(&acts);
+    posix_spawnattr_destroy(&attrs);
+    Nan::ThrowError("posix_spawn_file_actions_init failed.");
+    return 1;
+  }
+
+  if (
+    posix_spawnattr_setsigdefault(&attrs, &all_signals) ||
+    posix_spawnattr_setsigmask(&attrs, &oldmask) ||
+    posix_spawnattr_setflags(&attrs, POSIX_SPAWN_SETSIGDEF | POSIX_SPAWN_SETSIGMASK | POSIX_SPAWN_SETSID)
```

Comment:
Is POSIX_SPAWN_SETSID valid here? All I found is that it is not officially supported by any system (see https://www.austingroupbugs.net/view.php?id=1044 and https://stackoverflow.com/a/46828718/12548337).

In any case, this needs serious testing - whether the controlling terminal is correctly attached, and the first process forms its own process group (also see POSIX_SPAWN_RESETIDS andPOSIX_SPAWN_SETPGROUP). If the terminal gets not correctly attached as cterm, then we might have to double exec into a small `setsid` helper, as described in the link above.

Reply:
It appears that at least [glibc](https://man7.org/linux/man-pages/man3/posix_spawn.3.html) and macOS support SETSID - I'll add an availability check with fallback